incubator/hnc/internal/validators/hncconfig.go

```diff
@@ -141,6 +147,171 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
 	return allow("")
 }
 
+// objNameSet is a set of namespaced objects with object names as the key. It
+// also stores if the object has been checked for naming conflict.
+type objNameSet map[string]inNamespace
+type inNamespace struct {
+	namespace string
+	checked   bool
+}
+
+// conflicts stores object naming conflicts by GVK, object name and groups of
+// conflicting namespaces. The namespaces are grouped by the first ancestor.
+type conflicts map[schema.GroupVersionKind]map[string]affectedNamespaces
+type affectedNamespaces map[string][]string
+
+func (c *HNCConfig) checkForest(inst *api.HNCConfiguration) admission.Response {
+	c.Forest.Lock()
+	defer c.Forest.Unlock()
+
+	// Get types that are changed from other modes to "Propagate" mode.
+	gvks := c.getNewPropagateTypes(inst)
+
+	// Check if user-created objects would be overwritten by these mode changes.
+	conflicts := conflicts{}
+	for gvk, _ := range gvks {
+		c.checkConflictsForGVK(gvk, conflicts)
+	}
+
+	if len(conflicts) != 0 {
+		return deny(metav1.StatusReasonConflict, printMsgWithConflicts(conflicts))
+	}
+	return allow("")
+}
+
+// checkConflictsForGVK checks object naming conflicts for a specific type.
+func (c *HNCConfig) checkConflictsForGVK(gvk schema.GroupVersionKind, conflicts conflicts) {
+	// We will keep one object set for each type, because objects with the same
+	// name but different types won't affect each other.
+	obs := objNameSet{}
```

Comment:
I find this logic quite hard to reason about, because we process the namespaces in a random order but maintain state between visiting each one.

As an alternative, how about this: do a _top-down_ traversal with the only information being passed from parent to child being the list of objects and the namespaces that they're from (since not all source objects are propagated, especially with exceptions, there can be multiple ancestors with objects with the same name). E.g. something like this:

```
// ancestorObjects maps an object name to the ancestor namespace(s) in which it's defined
type ancestorObjects map[string][]string
func (a ancestorObjects) copy() ancestorObject {... makes a copy ...}

func (...) checkConflictsForGVK(gvk) []string {
  conflicts := []string{}
  for _, ns := range c.Forest.GetRoots() {
    conflicts = append(conflicts, c.checkConflictsForNS(gvk, ancestorObjects{}, ns)
  }
}

func (...) checkConflictsForNS(gvk, ao, ns) []string {
  conflicts := []string{}
  objs := ao.copy()
  for _, o := ns.GetSourceObjects(gvk) {
    onm := o.GetName()
    if len(objs[onm]) > 0 {
      // TODO: check if this is a real conflict
      conflicts = append(conflicts, fmt.Sprintf("Object %q in namespace %q overwrites object in %q", onm, <todo>, ns.GetName())
      continue
    }
    objs[onm] = append(objs[onm], ns.GetName())
  }
  for _, cns := ns.GetChildren() {
    conflicts = append(conflicts, r.checkConflictsForNS(gvk, objs, cns)
  }
  return cns
}
```